### PR TITLE
Use fake timer to get rid of sleeps in unit tests

### LINKED ISSUE #1157: Flaky test_rolling_with_local_store_writes fails due to timing
## Description
This test should have more reliable output or be disabled until we write better tests that cover this use case.

See: https://github.com/whylabs/whylogs/actions/runs/4506601154/jobs/7933562528?pr=1155

- [x] I have reviewed the [Guidelines for Contributing](https://github.com/whylabs/whylogs/blob/mainline/.github/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/whylabs/whylogs/blob/mainline/.github/CODE_OF_CONDUCT.md).

### PR DESCRIPTION
**Pull request by richard-rogers**
## Description

The rolling logger and a couple other tests used `time.sleep()`. This uses a faked `threading.Timer` that uses a simulated (single-threaded) schedule rather than real-time sleeping to speed up the tests. 

## Changes

- Use a fake timer instead of sleeping
- Adds a modified `LocalWriter` so profiles written in the same second don't collide

- [ ] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py [hunk 1]
```diff
@@ -1,11 +1,52 @@
...
+from threading import Timer
+from typing import Callable, List
 
 from whylogs.api.logger.experimental.multi_dataset_logger.time_util import (
     FunctionTimer,
     Schedule,
     TimeGranularity,
 )
 
+# A sequence of times get instantiated, so we need global state for them
```
Comment (richard-rogers) [inline python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py]:
```suggestion
# A sequence of timers get instantiated, so we need global state for them
```

File: python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py [hunk 2]
```diff
@@ -1,11 +1,52 @@
...
 from whylogs.api.logger.experimental.multi_dataset_logger.time_util import (
     FunctionTimer,
     Schedule,
     TimeGranularity,
 )
 
+# A sequence of timers get instantiated, so we need global state for them
+# to simulate universal time.
+
+_FAKE_TIMES: List[float] = []  # list of times (seconds) where the FakeTimer
```
Comment (richard-rogers) [inline python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py]:
```suggestion
_FAKE_TIMES: List[float] = []  # list of times (seconds) where the FakeTimer checks if it should execute
```

File: python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py [hunk 3]
```diff
@@ -1,11 +1,52 @@
...
     FunctionTimer,
     Schedule,
     TimeGranularity,
 )
 
+# A sequence of timers get instantiated, so we need global state for them
+# to simulate universal time.
+
+_FAKE_TIMES: List[float] = []  # list of times (seconds) where the FakeTimer checks if it should execute
+_TIME_INDEX: int = 0  # where are we in the lsit of times
```
Comment (richard-rogers) [inline python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py]:
```suggestion
_TIME_INDEX: int = 0  # where are we in the list of times
```

File: python/tests/api/logger/test_rolling.py [hunk 4]
```diff
@@ -4,20 +4,108 @@
...
 from whylogs.api.store.local_store import LocalStore
 from whylogs.api.store.query import DatasetIdQuery
+from whylogs.api.writer.local import LocalWriter
+from whylogs.api.writer.writer import Writable
 from whylogs.core.errors import BadConfigError
 from whylogs.core.schema import DatasetSchema
 from whylogs.core.segmentation_partition import segment_on_column
 
 
+# A sequence of times get instantiated, so we need global state for them
```
Comment (richard-rogers) [inline python/tests/api/logger/test_rolling.py]:
```suggestion
# A sequence of timers get instantiated, so we need global state for them
```

File: python/tests/api/logger/test_rolling.py [hunk 5]
```diff
@@ -4,20 +4,108 @@
...
+from whylogs.api.writer.writer import Writable
 from whylogs.core.errors import BadConfigError
 from whylogs.core.schema import DatasetSchema
 from whylogs.core.segmentation_partition import segment_on_column
 
 
+# A sequence of timers get instantiated, so we need global state for them
+# to simulate universal time.
+
+_FAKE_TIMES: List[float] = []  # list of times (seconds) where the FakeTimer
```
Comment (richard-rogers) [inline python/tests/api/logger/test_rolling.py]:
```suggestion
_FAKE_TIMES: List[float] = []  # list of times (seconds) where the FakeTimer checks if it should execute
```

File: python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py [hunk 6]
```diff
@@ -41,8 +82,11 @@ def inc(self) -> None:
             self.i += 1
 
     f = Foo()
-    t = FunctionTimer(Schedule(TimeGranularity.Second, 1), f.inc)
-    time.sleep(3)
+    global _FAKE_TIMES, _TIME_INDEX
+    _FAKE_TIMES = [0, 0.5, 1, 1.9, 2, 2.1, 3]
+    _TIME_INDEX = 0
+    t = FunctionTimer(Schedule(TimeGranularity.Second, 1), f.inc, FakeTimer)
+    # time.sleep(3)
```
Comment (jamie256) [inline python/tests/api/logger/experimental/multi_dataset_logger/test_time_util.py]: remove commented out code?

File: python/tests/api/logger/test_rolling.py [hunk 7]
```diff
@@ -1,22 +1,109 @@
...
+        now = _FAKE_TIMES[_TIME_INDEX]
+        if (now - self._prev_start_time) >= self._interval:
+            self._fn()
+            self._prev_start_time = now
+        else:
+            self.start()
+
+    def cancel(self) -> None:
+        global _TIME_INDEX
+        _TIME_INDEX = len(_FAKE_TIMES) + 1
```
Comment (jamie256) [inline python/tests/api/logger/test_rolling.py]: Why do we duplicate this test class?

File: python/tests/api/logger/test_rolling.py [hunk 8]
```diff
@@ -32,44 +119,66 @@ def test_closing(tmp_path: Any, lending_club_df: pd.DataFrame) -> None:
...
+    # start = time.time()
     with why.logger(mode="rolling", interval=1, when="S", base_name="test_base_name") as logger:
-        logger.append_writer("local", base_dir=tmp_path)
+        logger.append_writer(writer=TestWriter(base_dir=tmp_path))
         logger.log(df)
-        time.sleep(1.5)
+        _FAKE_TIMES = [x / 10.0 for x in range(16)]
+        _TIME_INDEX = 0
+        logger._scheduler._timer.start()
+        # time.sleep(1.5)
```
Comment (jamie256) [inline python/tests/api/logger/test_rolling.py]: remove the commented out sleep

**Review by jamie256 — COMMENTED** [review]
This is a lot better! Minor cleanup comments.

**Review by jamie256 — APPROVED** [review]
LGTM!

**Pull request merged**